n_beats/model_keras.py

```diff
@@ -0,0 +1,296 @@
+        plt.savefig('results/test/' + str(nb_epochs) + '.png')
+
+        print(get_accuracy(y_test, validation_predictions))
+
+
+def linear_space(length, fwd_looking=True):
+    if fwd_looking:
+        t = np.linspace(0, length - 1, length)
+    else:
+        t = np.linspace(-length, -1, length)
+    t = t / length
+    return t
+
+
+def seasonality_model(thetas, length, is_forecast):
+    p = thetas.shape[-1]
+    p1, p2 = (p // 2, p // 2) if p % 2 == 0 else (p // 2, p // 2 + 1)
+    t = linear_space(length, fwd_looking=is_forecast)
+    s1 = np.stack([np.cos(2 * np.pi * i * t) for i in range(p1)], axis=0)  # H/2-1
+    s2 = np.stack([np.sin(2 * np.pi * i * t) for i in range(p2)], axis=0)
+    S = np.concatenate([s1, s2], axis=0)
+    S = K.cast(S, np.float32)
+    return K.dot(thetas, S)
+
+
+def trend_model(thetas, length, is_forecast):
+    p = thetas.shape[-1]
+    t = linear_space(length, fwd_looking=is_forecast)
+    T = K.transpose(tf.stack([t ** i for i in range(p)], axis=0))
+    T = K.cast(T, np.float32)
+    return K.dot(thetas, K.transpose(T))
+
+
+def get_xtrain_circular(backcast_length, forecast_length):
+    # https://www.mcompetitions.unic.ac.cy/the-dataset/
+
+    x_train = np.array([]).reshape(0, backcast_length)
+    y_train = np.array([]).reshape(0, forecast_length)
+    x_tl = []
+    headers = True
+    filename = 'data/train/'
+    with open(filename + 'Daily-train.csv', "r") as file:
+        reader = csv.reader(file, delimiter=',')
+        for line in reader:
+            if not headers:
+                x_tl.append(line)
+            if headers:
+                headers = False
+    x_tl_tl = np.array(x_tl)
+    x_tl_tl = x_tl_tl[:, 1:]
+    for i in range(x_tl_tl.shape[0]):
+        if (len(x_tl_tl[i]) < backcast_length + forecast_length):
+            continue
+        time_series = np.array(x_tl_tl[i])
+        time_series = [float(s) for s in time_series if s != '']
+        time_series_cleaned = np.array(time_series)
+        time_series_cleaned_forlearning_x = np.zeros((1, backcast_length))
+        time_series_cleaned_forlearning_y = np.zeros((1, forecast_length))
+        j = np.random.randint(backcast_length, time_series_cleaned.shape[0] + 1 - forecast_length)
+        time_series_cleaned_forlearning_x[0, :] = time_series_cleaned[j - backcast_length: j]
+        time_series_cleaned_forlearning_y[0, :] = time_series_cleaned[j:j + forecast_length]
+        x_train = np.vstack((x_train, time_series_cleaned_forlearning_x))
+        y_train = np.vstack((y_train, time_series_cleaned_forlearning_y))
+
+    ts_min = np.min(x_train, axis = 1).reshape((x_train.shape[0], 1))
+    ts_max = np.max(x_train, axis = 1).reshape((x_train.shape[0], 1))
+    x_train = (x_train - ts_min) / (ts_max - ts_min + 1e-8)
+    y_train = (y_train - ts_min) / (ts_max - ts_min + 1e-8)
+    return x_train, y_train
+
+
+def get_xtest_circular(backcast_length, forecast_length):
```

Comment:
More nice to have: Could `get_xtrain_circular` and `get_xtest_circular` be just one function? But really not necessary here.

Reply:
I tried to factorize a bit